commit bbf188f3f8430c6491eadba38e9453d16398faf3 (refs/changes/44/3644/16)
Author: Rui Loura <rui@joyent.com>
Date:   2018-05-10T19:40:39+00:00 (1 year, 5 months ago)
    
    NET-390 portolan should support intra-DC cross vnet routing
    Reviewed by:
    Approved by:

diff --git a/lib/buckets.js b/lib/buckets.js
index 739267e..7c45737 100644
--- a/lib/buckets.js
+++ b/lib/buckets.js
@@ -86,6 +86,51 @@ var BUCKETS = {
                 version: 0
             }
         }
+    },
+
+    /*
+     * Key:  vnet_id,vlan_id,subnet,r_subnet
+     */
+    vnet_routes: {
+        desc: 'local to remote vnet mapping',
+        name: 'portolan_vnet_routes',
+        schema: {
+            index: {
+                vnet_id: {
+                    type: 'number'
+                },
+                vlan_id: {
+                    type: 'number'
+                },
+                subnet: {
+                    type: 'subnet'
+                },
+                r_dc_id: {
+                    type: 'number'
+                },
+                r_vnet_id: {
+                    type: 'number'
+                },
+                r_vlan_id: {
+                    type: 'number'
+                },
+                r_subnet: {
+                    type: 'subnet'
+                },
+                r_send_mac: {
+                    type: 'number'
+                },
+                version: {
+                    type: 'number'
+                },
+                deleted: {
+                    type: 'boolean'
+                }
+            },
+            options: {
+                version: 0
+            }
+        }
     }
 };
 var NAMES = {};
diff --git a/lib/common.js b/lib/common.js
index f7e6e41..415d54b 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -27,8 +27,19 @@ function vnetMacIPkey(ip, vid) {
     return [ip, vid].join(',');
 }
 
-
+/*
+ * Returns a key for the vnet_routes table
+ */
+function vnetRouteKey(opts) {
+    return [
+        opts.vnet_id,
+        opts.vlan_id,
+        opts.subnet,
+        opts.r_subnet
+    ].join(',');
+}
 
 module.exports = {
-    vnetMacIPkey: vnetMacIPkey
+    vnetMacIPkey: vnetMacIPkey,
+    vnetRouteKey: vnetRouteKey
 };
diff --git a/lib/consumer.js b/lib/consumer.js
index 84695be..91f116b 100644
--- a/lib/consumer.js
+++ b/lib/consumer.js
@@ -365,6 +365,50 @@ function logRm(opts, callback) {
     });
 }
 
+function vnetRouteLookup(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.object(opts.log, 'opts.log');
+    assert.object(opts.moray, 'opts.moray');
+    assert.number(opts.src_vnet_id, 'opts.src_vnet_id');
+    assert.number(opts.src_vlan_id, 'opts.src_vlan_id');
+
+    var vnet_id = opts.src_vnet_id;
+    var vlan_id = opts.src_vlan_id;
+
+    /*
+     * We get all local vnet & vlan remote mappings, and leave it up to
+     * Portolan to check if the src and dst IP's fall into the subnet and
+     * r_subnet.
+     */
+    var filter = fmt('&(vnet_id=%d)(vlan_id=%d)', vnet_id, vlan_id);
+    var searchOpts = {
+        limit: opts.limit,
+        sort: {
+            attribute: '_mtime'
+        }
+    };
+
+    var routes = [];
+
+    var req = opts.moray.findObjects(NAMES.vnet_routes, filter, searchOpts);
+
+    req.once('error', function searchError(err) {
+        opts.log.error(err, 'vnetRoute search error');
+        callback(err);
+    });
+
+    req.on('record', function onRecord(obj) {
+        if (obj.value.deleted) {
+            return;
+        }
+        routes.push(obj.value);
+    });
+
+    req.once('end', function onEnd() {
+        opts.log.debug({ routes: routes }, 'consumer.vnetRoute records');
+        callback(null, routes);
+    });
+}
 
 module.exports = {
     underlayLookup: underlayLookup,
@@ -374,5 +418,6 @@ module.exports = {
     vl2LookupCns: vl2LookupCns,
     vl3Lookup: vl3Lookup,
     logReq: logReq,
-    logRm: logRm
+    logRm: logRm,
+    vnetRouteLookup: vnetRouteLookup
 };
diff --git a/lib/producer.js b/lib/producer.js
index f59f406..0cbc159 100644
--- a/lib/producer.js
+++ b/lib/producer.js
@@ -48,6 +48,23 @@ function overlayMapRecord(opts) {
     };
 }
 
+function vnetRouteMapRecord(opts) {
+    return {
+        key: mod_common.vnetRouteKey(opts),
+        value: {
+            vnet_id: opts.vnet_id,
+            vlan_id: opts.vlan_id,
+            subnet: opts.subnet,
+            r_dc_id: opts.r_dc_id,
+            r_vnet_id: opts.r_vnet_id,
+            r_vlan_id: opts.r_vlan_id,
+            r_subnet: opts.r_subnet,
+            r_send_mac: opts.r_send_mac,
+            version: opts.version || VERSION,
+            deleted: opts.deleted || false
+        }
+    };
+}
 
 
 // --- Exports
@@ -244,6 +261,98 @@ function removeUnderlayMapping(opts, cb) {
     client.delObject(NAMES.underlay, key, cb);
 }
 
+function addVnetRouteMapping(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.object(opts.moray, 'opts.moray');
+    assert.number(opts.vnet_id, 'opts.vnet_id');
+    assert.number(opts.vlan_id, 'opts.vlan_id');
+    assert.string(opts.subnet, 'opts.subnet');
+    assert.number(opts.r_dc_id, 'opts.r_dc_id');
+    assert.number(opts.r_vnet_id, 'opts.r_vnet_id');
+    assert.number(opts.r_vlan_id, 'opts.r_vlan_id');
+    assert.string(opts.r_subnet, 'opts.r_subnet');
+    assert.number(opts.r_send_mac, 'opts.r_send_mac');
+    assert.optionalNumber(opts.version, 'opts.version');
+    assert.optionalBool(opts.deleted, 'opts.deleted');
+    assert.func(callback, 'callback');
+
+    var client = opts.moray;
+    var rec = vnetRouteMapRecord(opts);
+
+    client.putObject(NAMES.vnet_routes, rec.key, rec.value, callback);
+}
+
+/*
+ * XXX: UNUSED, for future use by NAPI.
+ */
+function updateVnetRouteMapping(opts, cb) {
+    assert.object(opts, 'opts');
+    assert.object(opts.moray, 'opts.moray');
+    assert.number(opts.vnet_id, 'opts.vnet_id');
+    assert.number(opts.vlan_id, 'opts.vlan_id');
+    assert.string(opts.subnet, 'opts.subnet');
+    assert.number(opts.r_dc_id, 'opts.r_dc_id');
+    assert.number(opts.r_vnet_id, 'opts.r_vnet_id');
+    assert.number(opts.r_vlan_id, 'opts.r_vlan_id');
+    assert.string(opts.r_subnet, 'opts.r_subnet');
+    assert.number(opts.r_send_mac, 'opts.r_send_mac');
+    assert.optionalNumber(opts.version, 'opts.version');
+    assert.optionalBool(opts.deleted, 'opts.deleted');
+    assert.func(cb, 'callback');
+
+    var client = opts.moray;
+    var key = mod_common.vnetRouteKey(opts);
+
+    client.getObject(NAMES.vnet_routes, key, function onUpdateGet(err, obj) {
+        if (err) {
+            cb(err);
+            return;
+        }
+
+        var rec_val = {
+            vnet_id: opts.vnet_id,
+            vlan_id: opts.vlan_id,
+            subnet: opts.subnet,
+            r_dc_id: opts.r_dc_id || obj.r_dc_id,
+            r_vnet_id: opts.r_vnet_id || obj.r_vnet_id,
+            r_vlan_id: opts.r_vlan_id || obj.r_vlan_id,
+            r_subnet: opts.r_subnet,
+            r_send_mac: opts.r_send_mac || obj.r_send_mac,
+            version: opts.version || obj.version || VERSION,
+            deleted: opts.deleted || obj.deleted || false
+        };
+
+        var putOpts = {
+            etag: obj._etag
+        };
+
+        client.putObject(NAMES.vnet_routes, key, rec_val, putOpts,
+            function onUpdatePut(putErr, data) {
+
+            if (putErr) {
+               cb(putErr);
+               return;
+            }
+
+            cb(null, data);
+        });
+    });
+}
+
+function removeVnetRouteMapping(opts, cb) {
+    assert.object(opts, 'opts');
+    assert.object(opts.moray, 'opts.moray');
+    assert.number(opts.vnet_id, 'opts.vnet_id');
+    assert.number(opts.vlan_id, 'opts.vlan_id');
+    assert.string(opts.subnet, 'opts.subnet');
+    assert.string(opts.r_subnet, 'opts.r_subnet');
+
+    var client = opts.moray;
+    var key = mod_common.vnetRouteKey(opts);
+
+    client.delObject(NAMES.vnet_routes, key, cb);
+}
+
 /*
  * Produces a moray-batchable set of SVP VL2 logs, indicating that a shootdown
  * is required. These are required when an IP:MAC mapping is deleted (which
@@ -326,6 +435,58 @@ function vl3CnEventBatch(opts) {
     return batch;
 }
 
+/*
+ * Produce a moray-batchable set of SVP ROUTE logs, indicating that a an
+ * overlay mapping needs to be updated or removed.  This is triggered when a
+ * network's attached networks array has been updated.
+ */
+function vnetRouteEventBatch(opts) {
+    assert.object(opts, 'opts');
+    assert.object(opts.vnetCns, 'opts.vnetCns');
+    assert.number(opts.src_vnet_id, 'opts.src_vnet_id');
+    assert.number(opts.dst_vnet_id, 'opts.dst_vnet_id');
+    assert.number(opts.dcid, 'opts.dcid');
+    assert.string(opts.srcip, 'opts.srcip');
+    assert.string(opts.dstip, 'opts.dstip');
+    assert.number(opts.src_vlan_id, 'opts.src_vlan_id');
+    assert.number(opts.dst_vlan_id, 'opts.src_vlan_id');
+    assert.number(opts.src_prefixlen, 'opts.src_prefixlen');
+    assert.number(opts.dst_prefixlen, 'opts.dst_prefixlen');
+    assert.optionalNumber(opts.version, 'opts.version');
+
+    var batch = [];
+    batch = opts.vnetCns.map(function vnetCnsMapOne(cn) {
+        var uuid = mod_uuid.v4();
+        return {
+            bucket: NAMES.net_events,
+            key: uuid,
+            operation: 'put',
+            options: {
+                etag: null
+            },
+            value: {
+                cn_uuid: cn,
+                vnet_id: opts.src_vnet_id,
+                version: opts.version || VERSION,
+                record: {
+                    type: 'SVP_LOG_ROUTE',
+                    src_vnet_id: opts.src_vnet_id,
+                    dst_vnet_id: opts.dst_vnet_id,
+                    dcid: opts.dcid,
+                    srcip: opts.srcip,
+                    dstip: opts.dstip,
+                    src_vlan_id: opts.src_vlan_id,
+                    dst_vlan_id: opts.dst_vlan_id,
+                    src_prefixlen: opts.src_prefixlen,
+                    dst_prefixlen: opts.dst_prefixlen
+                }
+            }
+        };
+    });
+
+    return batch;
+}
+
 module.exports = {
     addOverlayMapping: addOverlayMapping,
     overlayMappingBatch: overlayMappingBatch,
@@ -339,5 +500,10 @@ module.exports = {
     removeUnderlayMapping: removeUnderlayMapping,
 
     vl2CnEventBatch: vl2CnEventBatch,
-    vl3CnEventBatch: vl3CnEventBatch
+    vl3CnEventBatch: vl3CnEventBatch,
+    vnetRouteEventBatch: vnetRouteEventBatch,
+
+    addVnetRouteMapping: addVnetRouteMapping,
+    updateVnetRouteMapping: updateVnetRouteMapping,
+    removeVnetRouteMapping: removeVnetRouteMapping
 };
diff --git a/test/api.test.js b/test/api.test.js
index da4f8cd..c97232a 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -25,7 +25,7 @@ var test = require('tape');
 
 test('buckets exposed', function (t) {
     t.ok(mod_pm.buckets, 'buckets exposed');
-    [ 'mac_ip', 'underlay', 'net_events' ].forEach(function (b) {
+    [ 'mac_ip', 'underlay', 'net_events', 'vnet_routes' ].forEach(function (b) {
         t.ok(mod_pm.buckets[b], b + ' bucket');
     });
 
